infrastructure-playbooks/cephadm-adopt.yml

```diff
@@ -233,26 +233,38 @@
       run_once: true
       delegate_to: '{{ groups[mon_group_name][0] }}'
 
-    - name: generate cephadm ssh key
-      command: "{{ ceph_cmd }} cephadm generate-key"
-      changed_when: false
-      run_once: true
-      delegate_to: '{{ groups[mon_group_name][0] }}'
+    - name: configure root ssh key for remote connections
+      when: cephadm_ssh_user == 'root'
+      block:
+        - name: generate cephadm ssh key
+          command: "{{ ceph_cmd }} cephadm generate-key"
+          changed_when: false
+          run_once: true
+          delegate_to: '{{ groups[mon_group_name][0] }}'
 
-    - name: get the cephadm ssh pub key
-      command: "{{ ceph_cmd }} cephadm get-pub-key"
-      changed_when: false
-      run_once: true
-      register: cephadm_pubpkey
-      delegate_to: '{{ groups[mon_group_name][0] }}'
+        - name: get the cephadm ssh pub key
+          command: "{{ ceph_cmd }} cephadm get-pub-key"
+          changed_when: false
+          run_once: true
+          register: cephadm_pubpkey
+          delegate_to: '{{ groups[mon_group_name][0] }}'
+
+    - name: use custom user with sudo rights instead of root for remote connections
+      when: cephadm_ssh_user != 'root'
+      block:
+        - name: add custom user private key
+          command: "{{ ceph_cmd }} config-key set mgr/cephadm/ssh_identity_key -i /home/{{ cephadm_ssh_user }}/.ssh/id_rsa"
```

Comment:
I'm unsure about the fact we are hardcoding `/home/{{ cephadm_ssh_user }}/.ssh/id_rsa`

Reply:
Before executing this playbook on the OSP side we're running a new command on the undercloud which is able to:
1. generate the `cephadm_ssh_user` on the overcloud nodes
2. create a keypair and distribute it to the overcloud nodes.

Assuming we use 'ceph-admin' as cephadm ssh user, this means that at some point the nodes that should be enrolled to the ceph cluster will have an `id_rsa` in the /home directory and can ssh each other using the ceph-admin user.

Here my considerations:

1. Having the key (both private and public) already distributed to the Ceph nodes helps here as you don't have to copy the key on the Ceph nodes (otherwise ssh fails)

2. The command on L256 (as well as the other command on L258) should be delegated to the first mon for two reasons:
    a. `cephadm` is run from a mon (and not from the undercloud/ceph-ansible node);
    b. -i `/home/{{ cephadm_ssh_user }}/.ssh/{{ cephadm_ssh_prikey_name | default('id_rsa') }}` is present on the mon node (same thing for cephadm_ssh_pubkey_name)

Reply:
> 2. The command on L256 (as well as the other command on L258) should be delegated to the first mon for two reasons:
>    a. `cephadm` is run from a mon (and not from the undercloud/ceph-ansible node);

definitely, yes

>    b. -i `/home/{{ cephadm_ssh_user }}/.ssh/{{ cephadm_ssh_prikey_name | default('id_rsa') }}` is present on the mon node (same thing for cephadm_ssh_pubkey_name)

is there a scenario where `/home/{{ cephadm_ssh_user }}/.ssh/{{ cephadm_ssh_prikey_name | default('id_rsa') }}` wouldn't be present on mon node? (eg: if someone generates keys from the 'ansible node' only)

Reply:
IMO, the content of keys should be registered in a variable from the ansible node and then passed to `cephadm` via `stdin` so it works anyway